### PR TITLE
Adding smart content present as

### LINKED ISSUE #307: Allow meta title for Template Content Type Parameters and add Content Type Values
e.g.

``` xml
<property name="select" type="select">
    <meta>
        <title lang="de">Select</title>
        <title lang="en">Select</title>
     </meta>
     <params>
         <param name="option" value="1">
             <meta>
                 <title lang="de">Herr</title>
                 <title lang="en">Mr</title>
             </meta>
         </param>
         <param name="option" value="2">
             <meta>
                 <title lang="de">Frau</title>
                 <title lang="en">Mrs</title>
             </meta>
         </param>
    </params>
</property>

```

### PR DESCRIPTION
**Pull request by wachterjohannes**
__Tasks:__

- [x] test coverage
- [ ] gather feedback for my changes
- [x] added changelog line
- [x] smart-content save present as
- [x] translatable params
- [x] inject params to javascript

__Changed ContentTypes:__

* InternalLinks
* SmartContent
* Location
* MediaSelection
* Snippet
* TextEditor

__Informations:__

| Q             | A
| ------------- | ---
| Tests pass?   | yes
| Fixed tickets | fixes #307, fixes https://github.com/sulu-cmf/sulu-standard/issues/315
| BC Breaks     | https://github.com/sulu-cmf/sulu/commit/0497f0749efa2b2ddb6756d7956d637d86dee26c
| Doc           | https://github.com/sulu-cmf/docs/commit/a23434ccd6ef9d9340bf66204003ef910fd26ef1

### CODE REVIEW

**Comment by wachterjohannes** [comment]
@danrot ready to review! please test the refactored content types (written down in description)

File: UPGRADE.md [hunk 1]
```diff
@@ -2,6 +2,70 @@
 
 ## dev-develop
 
+### Content Types
+
+PropertyParameter are now able to hold metadata. Therefor the Interface has changed. Please check all your ContentTypes
```
Comment (danrot) [inline UPGRADE.md]: Therefore

File: UPGRADE.md [hunk 2]
```diff
@@ -2,6 +2,70 @@
...
+
+```twig
+As String:
+{{ params.name }}
+
+As Array or boolean:
+{{ params.name.value }}
+
+Get translated title:
+{{ params.getTitle('de') }}
```
Comment (danrot) [inline UPGRADE.md]: We agreed on having this only for translatable parameters, right?

**Comment by danrot** [comment]
For the future: please add information like "I also changed the order in the navigation" to the PR description :-)

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js [hunk 3]
```diff
@@ -26,7 +26,7 @@
  * @params {Integer} [options.preSelectedSortBy] array with id of the preselected sort-possibility
  * @params {String} [options.preSelectedSortMethod] Sort-method to begin with (asc or desc)
  * @params {Array} [options.presentAs] array of presentation-possibilities with id and name property
- * @params {Integer} [options.preSelectedPresentAs] array with id presentation-possibility to begin with
+ * @params {Integer} [options.preSelectedPresentAs] id presentation-possibility to begin with
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js]: id __of__ the default presentation __mode__ ?

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js [hunk 4]
```diff
@@ -129,7 +128,6 @@ define([], function() {
             hideCategories: false,
             hideTags: false,
             hideSortBy: false,
-            hidePresentAs: false,
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js]: I wouldn't remove this line, since it might be possible that we want to make this configureable.

File: src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig [hunk 5]
```diff
@@ -10,7 +10,7 @@
      data-form="true"
      data-aura-column-navigation-url="/admin/api/nodes{/dataSource}?tree=true{% if webspaceKey != '' %}&webspace={{ webspaceKey }}{% endif %}&language={{ languageCode }}&webspace-node=true"
      data-aura-hide-categories="true"
-     data-aura-hide-present-as="true"
+     data-aura-present-as='{{ parameterToSelect(params.present_as.value, userLocale)|json_encode }}'
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/views/Template/content-types/smart_content.html.twig]: Is that your own Twig extension? Would call it `parameter_to_select` nennen.

File: src/Sulu/Bundle/LocationBundle/Resources/views/Template/content-types/location.html.twig [hunk 6]
```diff
@@ -7,8 +7,8 @@
      data-type-instance-name="{{ id|raw }}"
      data-instance-name="{{ id|raw }}"
      data-form="true"
-     data-aura-countries="{{ params.countries|json_encode }}"
-     data-aura-map-providers="{{ params.mapProviders|json_encode }}"
+     data-aura-countries="{{ parameterToKeyValue(params.countries.value)|json_encode }}"
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/views/Template/content-types/location.html.twig]: Aren't you missing a locale here?

File: src/Sulu/Bundle/MediaBundle/Content/Types/MediaSelectionContentType.php [hunk 7]
```diff
@@ -44,25 +45,29 @@ public function __construct($mediaManager, $template)
...
+                array(
+                    'leftTop' => new PropertyParameter('leftTop', 'string', true),
+                    'top' => new PropertyParameter('top', 'string', true),
+                    'rightTop' => new PropertyParameter('rightTop', 'string', true),
+                    'left' => new PropertyParameter('left', 'string', true),
+                    'middle' => new PropertyParameter('middle', 'string', false),
+                    'right' => new PropertyParameter('right', 'string', true),
+                    'leftBottom' => new PropertyParameter('leftBottom', 'string', true),
+                    'bottom' => new PropertyParameter('bottom', 'string', true),
+                    'rightBottom' => new PropertyParameter('rightBottom', 'string', true)
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Content/Types/MediaSelectionContentType.php]: Storing true as a string feels quite strange... Shouldn't we make the type nullable, and only have collection and such types?

File: src/Sulu/Component/Content/PropertyParameter.php [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    private $metadata;
+
+    /**
+     * Constructor
+     * @param string $name
+     * @param string $type
+     * @param string|bool|array $value
+     * @param array $metadata
+     */
+    function __construct($name, $type, $value, $metadata = array())
```
Comment (danrot) [inline src/Sulu/Component/Content/PropertyParameter.php]: Wouldn't it be more comfortable to have the order `$name, $value, $type, $metadata`, and default type to null? So that it is only set if needed? I also would not use the type string, but for special stuff like collections.

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/smart-content/main.js]: i have changed the behavior to if no present as options are past it is hidden and vice versa!

File: UPGRADE.md (hunk 2 above)
Comment (wachterjohannes) [inline UPGRADE.md]: ou there is a bug! this should be `{{ params.name.getTitle('de') }}`

Comment (wachterjohannes) [inline UPGRADE.md]: and a better example is `{{ params.testParam }}, {{ params.testParam.name }}, {{ params.testParam.value }}, {{ params.testParam.getTitle('de') }}`

File: src/Sulu/Bundle/LocationBundle/Resources/views/Template/content-types/location.html.twig (hunk 6 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/LocationBundle/Resources/views/Template/content-types/location.html.twig]: yes thats right ... shouldn´t we do it in another PR? this is not really related to smart content. i have created a new issue https://github.com/sulu-cmf/sulu/issues/835

File: src/Sulu/Component/Content/PropertyParameter.php (hunk 8 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Content/PropertyParameter.php]: i fix this

File: src/Sulu/Bundle/MediaBundle/Content/Types/MediaSelectionContentType.php (hunk 7 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Content/Types/MediaSelectionContentType.php]: i will fix this

**Comment by wachterjohannes** [comment]
@danrot fixed everything!

File: UPGRADE.md (hunk 2 above)
Comment (danrot) [inline UPGRADE.md]: Doesn't seem fixed yet ;-) And what do you want to say with your examples?

Comment (wachterjohannes) [inline UPGRADE.md]: @danrot we have only translatable values now!

File: src/Sulu/Bundle/LocationBundle/Resources/views/Template/content-types/location.html.twig (hunk 6 above)
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/views/Template/content-types/location.html.twig]: That was a more general question, shouldn't the parameterToKeyValue take a localization as well?

**Comment by danrot** [comment]
Can you rename the `ContentExtension` to `ContentTwigExtension` for consistency?

**Comment by danrot** [comment]
The country selection in the location content type looks a bit weird:
image
